### PR TITLE
Add controllable warning about use of legacy settings.

### PR DESCRIPTION
**Pull request by sbc100**
This allows legacy settings warning to be controlled via the normal
means of setting `-Wno-legacy-settings` or `-Wlegacy-settings`.

### CODE REVIEW

**Comment by sbc100** [comment]
@juj I know you were part of the original discussion around how to message the use of legacy settings.  I kind of like this method, because it means that most users will see warnings, but users who want to support multiple emscriten versions at the same time with the same build flags and just add `-Wno-legacy-settings` to disable the warnings.

**Comment by juj** [comment]
I am still in favor of the concept that if there once existed an option, `-s FOO=0`/`-s FOO=1`, and if that option is removed in favor of always behaving the way that one of those options did, say, `-s FOO=1`, then emcc should without any extra hoops just be silent when `-s FOO=1` is passed (and error if the now unsupported `-s FOO=0` setting is passed).

Same goes for aliased settings - the old alias name should also be supported, along with the new canonical name.

I think that behavior is what should be considered sane and normal.

In `-s STRICT=1` build mode one can error&warn as much as one likes about removed options and aliases, if that is really desired. Btw such errors and warnings would be good to tell which version dropped the setting. The current warnings list reads

```js
var LEGACY_SETTINGS = [
  ['BINARYEN', 'WASM'],
  ['BINARYEN_ASYNC_COMPILATION', 'WASM_ASYNC_COMPILATION'],
  ['UNALIGNED_MEMORY', [0], 'forced unaligned memory not supported in fastcomp'],
  ['FORCE_ALIGNED_MEMORY', [0], 'forced aligned memory is not supported in fastcomp'],
  ['PGO', [0], 'pgo no longer supported'],
  ['QUANTUM_SIZE', [4], 'altering the QUANTUM_SIZE is not supported'],
  ['FUNCTION_POINTER_ALIGNMENT', [2], 'Starting from Emscripten 1.37.29, no longer available (https://github.com/emscripten-core/emscripten/pull/6091)'],
  ['BUILD_AS_SHARED_LIB', [0], 'Starting from Emscripten 1.38.16, no longer available (https://github.com/emscripten-core/emscripten/pull/7433)'],
  ['SAFE_SPLIT_MEMORY', [0], 'Starting from Emscripten 1.38.19, SAFE_SPLIT_MEMORY codegen is no longer available (https://github.com/emscripten-core/emscripten/pull/7465)'],
  ['SPLIT_MEMORY', [0], 'Starting from Emscripten 1.38.19, SPLIT_MEMORY codegen is no longer available (https://github.com/emscripten-core/emscripten/pull/7465)'],
  ['BINARYEN_METHOD', ['native-wasm'], 'Starting from Emscripten 1.38.23, Emscripten now always builds either to Wasm (-s WASM=1 - default), or to asm.js (-s WASM=0), other methods are not supported (https://github.com/emscripten-core/emscripten/pull/7836)'],
  ['PRECISE_I64_MATH', [1, 2], 'Starting from Emscripten 1.38.26, PRECISE_I64_MATH is always enabled (https://github.com/emscripten-core/emscripten/pull/7935)'],
  ['MEMFS_APPEND_TO_TYPED_ARRAYS', [1], 'Starting from Emscripten 1.38.26, MEMFS_APPEND_TO_TYPED_ARRAYS=0 is no longer supported. MEMFS no longer supports using JS arrays for file data (https://github.com/emscripten-core/emscripten/pull/7918)'],
  ['ERROR_ON_MISSING_LIBRARIES', [1], 'missing libraries are always an error now'],
  ['EMITTING_JS', [1], 'The new STANDALONE_WASM flag replaces this (replace EMITTING_JS=0 with STANDALONE_WASM=1)'],
  ['SKIP_STACK_IN_SMALL', [0, 1], 'SKIP_STACK_IN_SMALL is no longer needed as the backend can optimize it directly'],
  ['SAFE_STACK', [0], 'Replace SAFE_STACK=1 with STACK_OVERFLOW_CHECK=2'],
  ['MEMORY_GROWTH_STEP', 'MEMORY_GROWTH_LINEAR_STEP'],
  ['WASM_OBJECT_FILES', [1], 'Use -flto or -fto=thin instead'],
  ['TOTAL_MEMORY', 'INITIAL_MEMORY'],
  ['WASM_MEM_MAX', 'MAXIMUM_MEMORY'],
  ['BINARYEN_MEM_MAX', 'MAXIMUM_MEMORY'],
];
```
Note how the first error messages explicitly say "starting from Emscripten x.yy.z", and a link to the change, but the latter ones that have been added read a bit sloppy without reference information.

I think printing that reference information to the user is what is important, rather than adding noisy warnings. Such warnings will just prompt people to remove the use of those options, and when that "people" happens to be a build system author doing work for other ecosystems, they will unknowingly break backwards compatibility for other people, for no real benefit or fix to any real problem.

In summary, instead of `WarningManager.warn()`, it would be better to use `logger.debug()` in this PR.

**Comment by sbc100** [comment]
I agree its useful to have those very numbers in there.   However, I think its even more useful if people actually see the warning.   Putting this stuff behind EMCC_DEBUG=1 means that basically nobody will see it because it ends up buried in hundreds of lines of output.

I think that your use case is the exception.  I imagine most users who upgrade their SDK want to remove their legacy flags right away and support just one version.   The way I see it, users like yourself who maintain SDKs that support multiple emscripten versions should be savvy enough to add `-Wno-legacy-settings` if they want to continue to use legacy settings without being bombarded with warnings.

However, for the sake of compromise I propose that I disable this warning by default.  So only users of `-STRICT=1` or `-Wall` will see them  Does that sounds like a reasonable middle ground for now?

A bit context: The reason I would still like to one day turn this on by default is that I think its useful to inform users when they are using legacy/deprecated options.  It avoids confusion in the community.  For example many users might still be passing `-s BINARYEN=1` which leads to confusion.  What is that option for?  Should I be using it too?

**Comment by sbc100** [comment]
OK I disabled the warning by default.

**Comment by sbc100** [comment]
OK, this change should be uncontroversial now.

**Comment by juj** [comment]
> I think that your use case is the exception. I imagine most users who upgrade their SDK want to remove their legacy flags right away and support just one version.

I think this might highlight the misunderstanding. This is not about users who have a singular codebase and a single version of compiler they want to use. Imagining that compiler users are like that would be a simplification. This is about library/tool/engine/buildsystem projects where there is a collaboration of more than one person or team that takes place. Imagine e.g. SDL2, Urho3D, Bgfx or any other library project. One team provides the engine and the build system, but then other people come along and use that library, along with the build system it provides. The users of the libraries/build systems supply the compiler, and libray/build system authors have created the interaction with the compiler flags to instruct how that 3rd party piece of code should be built. Large projects can be comprised of a number of these kind of building blocks.

There is a large ecosystem of open source software users out there. This is not me from my "special esoteric build system plumbing" perspective, but from close to a decade of watching the Emscripten library ecosystem perspective.

Removing options that work perfectly well, just for the sake of "I did not like it for the hygiene" have actual "spent a day or two finding the compiler version that can build both libX and libY successfully" consequences to people. Let's avoid bidirectional breakage points when they give no benefit.

Warning about this is about the same as erroring, since we are talking about a collaboration of multiple teams/software projects. If one project team sees this kind of useless warning, they are prompted to bump the minimum needed compiler version, and then every user of that project has lost something of value for no gain. (just to appease the hygiene needs of compiler author)

I still think that this kind of thing should be a -s STRICT=1 and/or a logger.debug() print. Having it be a -Wall thing diminishes the value of -Wall, because it will begin to warn about things that are non-issues.

The fact that it feels non-hygienic as a compiler author to have to keep supporting e.g. `-s BINARYEN=1` or `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` - it is just a feeling. Please try to let it go!

**Comment by hujiajie** [comment]
> I imagine most users who upgrade their SDK want to remove their legacy flags right away and support just one version.

I'm one of those who prefer to get rid of legacy flags in my projects, but at the same time I'd like to support a few recent toolchain releases, so I tend to check the compiler version in my configure script/CMakeLists.txt and add the right flags where necessary.

Also, when building a program, I wish the command line could be similar enough to a normal toolchain, without using many Emscripten-specific flags. It seems that the Clang flag [-Wunused-command-line-arguments](https://clang.llvm.org/docs/DiagnosticsReference.html#wunused-command-line-argument) can already complain about similar situations, so is it possible to extend this in emcc?

**Comment by sbc100** [comment]
OK, while I generally prefer to move faster towards deprecation and removal it not a hill a I need to die on.

I moved this warning out of `-Wall`.  It can still be enabled explicitly and is enabled via `-pedantic`.  I hope that is reasonable.

@hujiajie good idea about `unused-command-line-argument`.  It might make sense to use for settings that no longer exist and are completely ignored.   Not all legacy settings fall into that category though.

**Comment by hujiajie** [comment]
> It might make sense to use for settings that no longer exist and are completely ignored. Not all legacy settings fall into that category though.

Thanks, it makes sense to me. But now I wonder how it's related with `-pedantic`. My impression is that `-pedantic` is for issues in the source code only.

**Comment by sbc100** [comment]
I've been working on making warning generated by emscripten internally controllable via the normal gcc/clang-style compiler flags.    `-Wall`, `-Werror`, `-Wextra`, `-Wpedantic` control these warnings just like the do warnings in clang. 

Do you think this would make more sense behind `-Wextra`?  Or only ever show it when a user passes `-Wlegacy-settings` explicitly?

**Comment by sbc100** [comment]
OK, I removed it from `-pedantic` too

**Comment by hujiajie** [comment]
@sbc100 I have no strong preference on whether `-Wextra` should imply `-Wlegacy-settings`.

**Comment by sbc100** [comment]
I decided to leave it as opt-in to keep this change simple.

**Comment by juj** [comment]
Thanks for accommodating, that is much appreciated!

**Comment by sbc100** [comment]
Can I get an LG on this?

File: emcc.py [hunk 1]
```diff
@@ -412,9 +409,14 @@ def standardize_setting_change(key, value):
     if key in shared.Settings.internal_settings:
       exit_with_error('%s is an internal setting and cannot be set from command line', key)
 
+    if key in shared.Settings.legacy_settings and key in shared.Settings.alt_names:
+      actual_key = shared.Settings.alt_names[key]
+    else:
+      actual_key = key
+
```
Comment (kripken) [inline emcc.py]: `key` is used in a few places lower down, and not `actual_key` (like `if key == 'EXPORTED_FUNCTIONS'`). that seems to be a risk for bugs. can we do `key = actual_key` or not even have `actual_key`? (alternatively, doing the aliasing in the standardize function like before achieves the same)

File: tools/shared.py [hunk 2]
```diff
@@ -827,6 +827,8 @@ def warn(warning_type, message, *args):
 WarningManager.add_warning('separate-asm')
 WarningManager.add_warning('almost-asm')
 WarningManager.add_warning('invalid-input')
+# Don't show lecacy settings warnings by default
```
Comment (kripken) [inline tools/shared.py]: `lecacy => legacy`

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py]: Ah yes.   Fixed.

**Comment by sbc100** [comment]
Should be good now, PTAL
